### PR TITLE
Add the proc size for streams

### PR DESCRIPTION
**Pull request by data-man**

### CODE REVIEW

**Comment by Araq** [comment]
Thanks a lot, it's pretty nice. However, it needs to be clearly documented what ``size`` means, "total size" (probably not) or "remaining size"?

**Comment by data-man** [comment]
How about the ```current size```?

**Comment by euantorano** [comment]
I think `len` would make more sense than `size` personally. `len` is already defined for other types like `string`, `seq`, etc.

**Comment by data-man** [comment]
@euantorano 
```len``` can be added as alias.
I focused on file and memfile in the first place.

**Comment by euantorano** [comment]
Fair enough, I'm happy either way.

**Comment by data-man** [comment]
I've one question: it's possible to remove the ```ReadIOEffect``` tag for StringStream?

**Comment by Araq** [comment]
> I've one question: it's possible to remove the ReadIOEffect tag for StringStream?

No.

**Comment by Araq** [comment]
> Up for discussion: len should return an Option[int].

Then it shouldn't be named ``len``. Also, the streams deal with bytes and https://nim-lang.org/docs/apis.html argues for the name ``size``.

**Comment by dom96** [comment]
Okay, fine ``size`` is fine.

Another idea: raise an exception when ``size`` is not implemented.

**Comment by Araq** [comment]
> Another idea: raise an exception when size is not implemented.

Ah, I see what you mean. ``Option[Natural]`` seems better then indeed and we should specialize it to only take a single word in memory... It's a "size *hint*". Also we might call it that way, ``sizeHint``.

**Comment by data-man** [comment]
> raise an exception when size is not implemented.

Some stream can be unlimited: ```atEnd``` always returns ```false``` and ```size``` always returns ```-1```.

**Comment by Varriount** [comment]
I second `sizeHint` and Option[Natural].

**Comment by data-man** [comment]
I disagree.
What for to complicate the simple and effective interface?
And why then Option is not used for other procs of Streams (read*, peek*)?

**Comment by bluenote10** [comment]
> What for to complicate the simple and effective interface?

Type safety.

**Comment by data-man** [comment]
@Araq what about this PR?

**Comment by Araq** [comment]
> @Araq what about this PR?

If you don't want the ``Option`` thing, at least name it ``sizeHint`` please.

**Comment by data-man** [comment]
```sizeHint``` it's a bad name. Is a hint of what?
File, memfile or string has exact size (or len), non hint. Any other stream should implement ```sizeImpl```.

**Comment by Araq** [comment]
Ok, then name it ``size`` but make it return ``Option[Natural]``

**Comment by data-man** [comment]
I don't see any advantage from the Option.
Well I prefer raise IOError exception with 'sizeImpl not implemented' message.

**Comment by Araq** [comment]
Yes, and I prefer the Option[Natural] over an exception. In this case. Otherwise you swap the streams implementation and the code starts to fail because you forgot to handle the "IO error".

**Comment by data-man** [comment]
So, with this PR as is:
```nim
var s = newStringStream("123")
let size = s.size
if size == -1:
  echo "size isn't implemented!"
else:
  echo size
```
With ```Option[Natural]```:
```nim
var s = newStringStream("123")
let size = s.size
if size.isNone:
  echo "size isn't implemented!"
else:
  echo size.get(0) # for e.g.
```
I don't see any advantage from the Option. :-)

**Comment by dom96** [comment]
> I don't see any advantage from the Option. :-)

The advantage is that you are forced to handle the ``none`` case. And it's also clear how to handle it: ``-1`` could mean anything.

**Comment by bluenote10** [comment]
> I don't see any advantage from the Option. :-)

You have never encountered the bug potential if users aren't aware that a "size" can be negative?

For instance, who would expect that something as innocent as `let s = newSeq[int](stream.sizeHint)` can fail with `Error: unhandled exception: value out of range: -1 [RangeError]`? Users should be made ware that `sizeHint` may not return a valid size and the compiler can help with that at compile time.

**Comment by data-man** [comment]
@dom96 
> -1 could mean anything.

No, the meaning of -1 is documented **very well**.
You haven't noticed it or don't want to notice?

@bluenote10 

strutils.find returns -1
It's safe for strings and ok for you?
grep ```return -1``` in stdlib dir, please.

**Comment by data-man** [comment]
The special meaning of -1 has been removed for the Universe happiness.

**Comment by Araq** [comment]
You still don't use ``Option[Natural]``.

**Comment by data-man** [comment]
I can't understand why you want it.
And why is there no such claim to getPosition and other stream's procs?

**Comment by zah** [comment]
@data-man, the idea is that `Option[SomeRangeType]` can be optimised to store only a single value (if you store a bit value that's outside of the range, this can be used to indicate the "no value" case).

An invalid value for any range may be `int(low(RangeType)) - 1`. This has the nice property of being `-1` for the `Natural` type.

**Comment by data-man** [comment]
@zah 
> the idea is
> TL;DR

This idea is absolutely not needed in this case (for last commit).

**Comment by data-man** [comment]
@Araq 

I found the beatiful issue! :-)
Some your opinions:
> But I don't like Option[T], it's messy to work with.
> Option[T] is annoying
> I still wouldn't like Option[T]
> And more...

I implemented the size for all stdlib's streams.
If someone has their own stream, then their implementation for this type must be implemented.

**Comment by data-man** [comment]
@dom96 
Please remove "requested changes" if name `size` is ok for you.

**Comment by arnetheduck** [comment]
@zah if you return `int(low(RangeType)) - 1`, which presumably is implementation-specific, what should the calling code do / compare with?

**Comment by zah** [comment]
@arnetheduck, this value is used only behind the scenes. From the perspective of the calling code, `isNone` returns true in such situations, so it's illegal to try to obtain the value and use it.

**Comment by data-man** [comment]
@Araq 

PTAL :)

**Comment by Araq** [comment]
I changed my mind, streams have no size. The most common streams come from sockets which lack this piece of information.
